pymc/distributions/mixture.py

```diff
@@ -112,454 +132,211 @@ class Mixture(Distribution):
+            ):
+                raise ValueError(
+                    f"Component dist must be a distribution created via the `.dist()` API, got {type(dist)}"
                 )
-            )
-        shape = kwargs.pop("shape", ())
+            check_dist_not_registered(dist)
+            components_ndim.add(dist.ndim)
+            components_ndim_supp.add(dist.owner.op.ndim_supp)
 
-        self.w = w = at.as_tensor_variable(w)
-        self.comp_dists = comp_dists
+        if len(components_ndim) > 1:
+            raise ValueError(
+                f"Mixture components must all have the same dimensionality, got {components_ndim}"
+            )
 
-        defaults = kwargs.pop("defaults", [])
+        if len(components_ndim_supp) > 1:
+            raise ValueError(
+                f"Mixture components must all have the same support dimensionality, got {components_ndim_supp}"
+            )
 
-        if all_discrete(comp_dists):
-            default_dtype = _conversion_map[aesara.config.floatX]
+        w = at.as_tensor_variable(w)
+
+        # ShapeWarning does not make sense here
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", ShapeWarning)
+            return super().dist([w, *comp_dists], **kwargs)
+
+    @classmethod
+    def rv_op(cls, weights, *components, size=None, rngs=None):
+        # Update rngs if provided
+        if rngs is not None:
+            *components_rngs, mix_indexes_rng = rngs
+            new_components = []
+            assert len(components) == len(components_rngs)
+            for component, component_rng in zip(components, components_rngs):
+                component_node = component.owner
+                old_rng, *inputs = component_node.inputs
+                new_components.append(
+                    component_node.op.make_node(component_rng, *inputs).default_output()
+                )
+            components = new_components
         else:
-            default_dtype = aesara.config.floatX
-
-            try:
-                self.mean = (w * self._comp_means()).sum(axis=-1)
-
-                if "mean" not in defaults:
-                    defaults.append("mean")
-            except AttributeError:
-                pass
-        dtype = kwargs.pop("dtype", default_dtype)
-
-        try:
-            if isinstance(comp_dists, Distribution):
-                comp_mode_logps = comp_dists.logp(comp_dists.mode)
-            else:
-                comp_mode_logps = at.stack([cd.logp(cd.mode) for cd in comp_dists])
+            # Create new rng for the choices internal RV
+            mix_indexes_rng = aesara.shared(np.random.default_rng())
+
+        # Extract support and replication ndims from components and weights
+        component_ndim_supp = components[0].owner.op.ndim_supp
+        component_ndim_reps = components[0].ndim - component_ndim_supp
+        weights_ndim_reps = max(0, weights.ndim - component_ndim_reps - 1)
```

Comment:
We could instead change the size of the `Mixture` so that there are never batched dimensions coming exclusively from the weights. Perhaps issue a UserWarning before doing this...